File: src/IoT.java
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class IoT {
    private volatile float temperature = 20.0f;
    private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);

    public IoT() {
        scheduler.scheduleAtFixedRate(this::fetchWeather, 0, 60, TimeUnit.SECONDS);
    }

    public void shutdown() {
        scheduler.shutdown();
        try {
            if (!scheduler.awaitTermination(5, TimeUnit.SECONDS)) {
                scheduler.shutdownNow();
            }
        } catch (InterruptedException e) {
            scheduler.shutdownNow();
            Thread.currentThread().interrupt();
        }
    }

    private void fetchWeather() {
        try {
            HttpClient client = HttpClient.newHttpClient();
            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create("https://api.open-meteo.com/v1/forecast?latitude=35.6895&longitude=139.6917&current_weather=true"))
                .build();

            client.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                .thenApply(HttpResponse::body)
                .thenAccept(this::parseTemperature)
                .exceptionally(e -> {
                    System.err.println("Failed to fetch weather: " + e.getMessage());
                    return null;
                });
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void parseTemperature(String json) {
        try {
            // Simple regex parsing to avoid external dependencies like Jackson/Gson
            Pattern pattern = Pattern.compile("\"temperature\":([0-9.-]+)");
            Matcher matcher = pattern.matcher(json);
            if (matcher.find()) {
                float newTemp = Float.parseFloat(matcher.group(1));
                // Sanity check: Tokyo temperature should be reasonable (-20 to 45°C)
                if (newTemp >= -20.0f && newTemp <= 45.0f) {
                    this.temperature = newTemp;
                    System.out.println("Updated temperature: " + this.temperature + "°C");
                } else {
                    System.err.println("Invalid temperature value: " + newTemp);
                }
            } else {
                System.err.println("Could not parse temperature from response");
            }
        } catch (NumberFormatException e) {
            System.err.println("Error parsing temperature value: " + e.getMessage());
        }
    }

    public float getTemperature() {
        return temperature;
    }
}

File: src/iot.rs
use reqwest::blocking::Client;
use serde::Deserialize;
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;

#[derive(Deserialize, Debug)]
struct WeatherResponse {
    current_weather: CurrentWeather,
}

#[derive(Deserialize, Debug)]
struct CurrentWeather {
    temperature: f32,
}

pub struct WeatherService {
    pub temperature: Arc<Mutex<f32>>,
}

impl WeatherService {
    pub fn new() -> Self {
        let temperature = Arc::new(Mutex::new(20.0)); // Default temp
        let temp_clone = temperature.clone();

        // Spawn a thread to fetch weather periodically
        thread::spawn(move || {
            let client = Client::new();
            loop {
                // Fetch for Tokyo (default)
                let url = "https://api.open-meteo.com/v1/forecast?latitude=35.6895&longitude=139.6917&current_weather=true";
                
                match client.get(url).send() {
                    Ok(resp) => {
                        if let Ok(weather_data) = resp.json::<WeatherResponse>() {
                            let mut temp = temp_clone.lock().unwrap();
                            *temp = weather_data.current_weather.temperature;
                            println!("Updated temperature: {} C", *temp);
                        }
                    }
                    Err(e) => {
                        eprintln!("Failed to fetch weather: {}", e);
                    }
                }

                // Wait for 60 seconds before next fetch
                thread::sleep(Duration::from_secs(60));
            }
        });

        WeatherService { temperature }
    }

    pub fn get_temp(&self) -> f32 {
        *self.temperature.lock().unwrap()
    }
}
